### ISSUE TITLE
`pulumi up [url]`

### ISSUE THREAD

**Issue by mehzer** [pulumi/pulumi#1504]
Was thinking about `hello-javascript-aws`, and packages generally.

What might be nice is a `search` function, and even maybe a parameterized `up` function for conveniently building packages.

**SEARCH**
```
pulumi search <criteria> [OPTIONS --lang js|ts|py|go]
Returns matching package name,description,author etc
```

**UP**
```
pulumi up <package> [OPTIONS --cloud aws|azu|gcp --env foo=bar]
```
Stands up a package in desired cloud, passing foo as bar. This could apply to only /cloud -based packages (or the package description could describe the constraints).

Example use (e.g. onboarding):

```
pulumi up @pulumi/hello-world --cloud AWS --env name=Marc
```
Would produce a functioning lambda (in JavaScript on AWS in this instance) that returns "Hello Marc"

**Comment by lukehoban** [comment]
> Since it's come up in discussion, for the pulumi up part, I think of this as distinctly different from new. I think of the above as triggering a workflow where you do not need to have the project source code locally (i.e., no need to run extra git clone commands). This makes the onboarding as smooth as possible.

I'll admit I actually like the idea of this being more like the existing `pulumi new` and giving you the code and an initialized stack locally.  This makes it totally trivial to take the next step of engagement and tweak a config setting or make a small code change.  This bridges between the "get an app stood up right away" and "how to I take the next step to customize this".  I also think it can be fairly low friction so that it adds value without watering down the case of "I just want to run it".

Perhaps it's fine to just have separate commands for both of these scenarios - but I think a mode that doesn't give you the code will require thinking through other aspects of how the workflow goes when you want to take another step: how do you destroy?  how do you update some configuration?  what happens if you ctrl-c in the middle of the deployment?  what happens if you don't have the right ambient credentials for AWS/Kube/etc.?

> As part of this, we'll need to think about whether we double down on how new decides to manage configuration, defaults, and packages, since it was a little bespoke and not general purpose. I still wish we could figure out a way to make #1052 "real" so that packages can self-advertise their config in metadata, along with the proposed default values.

Agreed.  I'd love to figure out #1052 here.  We'll have to decide though whether to hold up adding this richer `pulumi new` on figuring that out, or whether to light it up based on the existing [`.pulumi.template.yaml` model](https://github.com/pulumi/templates/blob/master/templates/hello-aws-javascript/.pulumi.template.yaml) for now and then work on #1052 in parallel.

**Comment by joeduffy** [comment]
> giving you the code and an initialized stack locally.

I definitely get the appeal, which is why I did suggest supporting both.  I just think this transitions from "I just want to run an application" to "I now need to manage a codebase", which is a very hard cliff.  I think it demos far better and gets you engaged much more quickly if you can just stand up and tear the thing down.  @ellismg can correct me if I'm wrong, but I believe the `--stack` flag would let us keep the full set of functionality we currently have, even without needing a project workspace.

Again, I agree `new` is super interesting too, I'd just hate to see us walk away from something that could be so much simpler for the simple scenarios (the place we're weak on right now).  We need to shift into a mindset of "get cloud to the cloud" with as few intermediary steps as possible, especially as we invest in workflows (where I think the need to configure and setup CI in between is going to also be a source of friction, despite being the "right" way to do it, especially for real world cases).

**Comment by justinvp** [comment]
I'm going to use this issue to track the `pulumi up` work items. This depends on some foundational work in `pulumi new` tracked by #1544.

- [ ] Investigate implications of using `up` with URL arg (how it impacts our "model" and other commands)
- [ ] Run preview by default (suppress output?)
- [ ] Prompt on failure with default option to rollback
- [ ] Verify great E2E experience for a few canonical apps (including one for K8s and one for AWS at least)
- [ ] Ensure we have a story for updates after an initial `up`

**Comment by justinvp** [comment]
Fixed by #1727. There are some follow-up items for the next milestone (#1764 and #1765).
